### PR TITLE
Create the infrastructure to allow testing changes across repos.

### PR DESCRIPTION
**Pull request by AlexGhiondea**
This is going to be driven by the localpublish.props file at the source root
of the repo. By default this file does not exist and it is not tracked by git.

You can create the file with contents similar to this to enable the repo that
is currently built from publishing to a local path. This enables the scenario
when one package has a dependency to the other and you want to test your
changes locally.

If the file is present the following are going to happen:
- The package folder in the target repo is going to be removed
- A packages.config file is going to be created for the package that we are
installing. This is required to bypass the resolving of dependencies that is
triggered by installing a packaged referenced by name.
- A nuget install command is issued to install the package in the target repo

Structure of localpublish.props file:
<Project ToolsVersion="12.0" DefaultTargets="Build"
         xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup>
    <Packages Include="<directory of the package created by the repo>">
      <PackageName><name of the package></PackageName>
      <PackageVersion><version, including -prerelease></PackageVersion>
      <InstallLocation><Location in the other repo where the package
      					should be installed</InstallLocation>
    </Packages>
</Project>

### CODE REVIEW

File: src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets [hunk 1]
```diff
@@ -76,4 +76,41 @@
...
+  
+  <!-- Publish the packages locally -->
+  <Target Name="LocalPackagePublish" Condition="'@(Packages)'!=''"
+          Inputs="%(Packages.Identity)\%(Packages.PackageName).%(Packages.PackageVersion).nupkg;%(Packages.InstallLocation)\%(Packages.PackageName).%(Packages.PackageVersion)" 
+          Outputs="%(Packages.InstallLocation)\localpublish.%(Packages.PackageName).%(Packages.PackageVersion).txt" >
+    <!--AfterTargets="BuildNuGetPackages"-->
+    
+    <!-- Generate the packages.config -->
+    <ItemGroup>
+      <Line Include="&lt;?xml version=&quot;1.0&quot; encoding=&quot;utf-8&quot;?&gt;" />
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: You should make this more specific. Something like LocalPackageConfigLine

Comment (AlexGhiondea) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: Done.

**Comment by AlexGhiondea** [comment]
I will rebase before I commit.

File: src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets [hunk 2]
```diff
@@ -18,6 +18,10 @@
     <NuGetOutputDirectory>&quot;$(PackagesOutDir.TrimEnd('\'))&quot;</NuGetOutputDirectory>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <LocalPublishPropsFile Condition="'$(LocalPublishPropsFile)'==''">$(SourceDir)\localpublish.props</LocalPublishPropsFile>
```
Comment (AlexGhiondea) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: If we don't want this to be $(SourceDir), what should it be? $(BinDir)? Something else?

Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: $(PackagesDir) is what I would use in corefx. if that is not set in coreclr we can probably just set it there to be consistent.

File: .gitignore [hunk 3]
```diff
@@ -14,4 +14,5 @@ packages
 Session.vim
 .netrwhist
 *~
+localpublish.props
```
Comment (weshaggard) [inline .gitignore]: Once you switch to something other than $(SourcesDir) you can remove this.

File: src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets [hunk 4]
```diff
@@ -76,4 +80,47 @@
...
+    <WriteLinesToFile File="$(RestoreConfig)" Lines="@(LocalPackageConfigLine)" Overwrite="true" />
+    <Message Text="Generated $(RestoreConfig) for '%(Packages.PackageName).%(Packages.PackageVersion).nupkg'" />
+    
+    <PropertyGroup>
+      <PackageSources>@(Packages, ';')</PackageSources>
+    </PropertyGroup>
+    <Message Text="Generated to package source to be '$(PackageSources)'"/>
+
+    <!-- Remove existing directories -->
+    <RemoveDir Directories="%(Packages.InstallLocation)\%(Packages.PackageName).%(Packages.PackageVersion)" />
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: We may want to make this more general at some point perhaps adding a * at the end. However lets stick with this one for now.

File: src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets [hunk 5]
```diff
@@ -76,4 +80,47 @@
...
+    <Exec Command="$(NuGetExe) install $(RestoreConfig) -Source $(PackageSources) -NoCache -Prerelease -OutputDirectory %(Packages.InstallLocation)" />
+
+    <!-- Write a file in the packages folder to let the user know this is not the original package -->
+    <ItemGroup>
+      <NotOriginalPackages Include="The package was restored from $(PackageSources)" />
+    </ItemGroup>
+    <WriteLinesToFile File="%(Packages.InstallLocation)\%(Packages.PackageName).%(Packages.PackageVersion)\MODIFIED.txt" Lines="@(NotOriginalPackages)" Overwrite="true" />
+
+
+    <Touch Files="%(Packages.InstallLocation)\localpublish.%(Packages.PackageName).%(Packages.PackageVersion).txt" AlwaysCreate="true" />
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: Why do we need to touch this file why not use the modified.txt instead?

File: src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets [hunk 6]
```diff
@@ -76,4 +80,47 @@
       <_TempPackageVersion />
     </PropertyGroup>
   </Target>
+
+  <Import Condition="Exists('$(LocalPublishPropsFile)')"  Project="$(LocalPublishPropsFile)"/>
+  
+  <!-- Publish the packages locally -->
+  <Target Name="LocalPackagePublish" Condition="'@(Packages)'!=''"
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: I actually think that Packages may conflict with other things we should consider a mores specific name. Perhaps LocalPackages.

File: src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets (hunk 2 above)
Comment (AlexGhiondea) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: I will make this $(PackagesDir) and will update the CoreCLR repo to have this variable

File: src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets (hunk 6 above)
Comment (AlexGhiondea) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: Makes sense. Updated.

File: src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets (hunk 5 above)
Comment (AlexGhiondea) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: I added the modified.txt later -- I will get rid of the other file.

File: .gitignore [hunk 7]
```diff
@@ -14,4 +14,3 @@ packages
 Session.vim
 .netrwhist
 *~
-
```
Comment (weshaggard) [inline .gitignore]: I assume you will undo the changes to this file when you squash.

File: src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets [hunk 8]
```diff
@@ -76,4 +80,44 @@
...
+      <RestoreConfig>%(LocalPackages.InstallLocation)\packages.%(LocalPackages.PackageName).config</RestoreConfig>
+    </PropertyGroup>
+    
+    <WriteLinesToFile File="$(RestoreConfig)" Lines="@(LocalPackageConfigLine)" Overwrite="true" />
+    <Message Text="Generated $(RestoreConfig) for '%(LocalPackages.PackageName).%(LocalPackages.PackageVersion).nupkg'" />
+    
+    <PropertyGroup>
+      <PackageSources>@(LocalPackages, ';')</PackageSources>
+    </PropertyGroup>
+    <Message Text="Generated to package source to be '$(LocalPackagesources)'"/>
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: You updated a few places to LocalPackagesources, I assume from a bad find/replace but the property group above is still called PackageSources.

File: src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets [hunk 9]
```diff
@@ -76,4 +80,44 @@
...
+
+    <!-- Remove existing directories -->
+    <RemoveDir Directories="%(LocalPackages.InstallLocation)\%(LocalPackages.PackageName).%(LocalPackages.PackageVersion)" />
+
+    <!-- Restoring the packages -->
+    <Exec Command="$(NuGetExe) install $(RestoreConfig) -Source $(LocalPackagesources) -NoCache -Prerelease -OutputDirectory %(LocalPackages.InstallLocation)" />
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: Should be PackageSources

File: src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets [hunk 10]
```diff
@@ -76,4 +80,44 @@
...
+
+    <!-- Write a file in the packages folder to let the user know this is not the original package -->
+    <ItemGroup>
+      <NotOriginalPackages Include="The package was restored from $(LocalPackagesources)" />
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: Should be PackageSources

**Comment by weshaggard** [comment]
@AlexGhiondea other than the last few comments and a squash the changes look good to me.

**Comment by jhendrixMSFT** [comment]
test this please

File: src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets [hunk 11]
```diff
@@ -76,4 +80,44 @@
...
+      <RestoreConfig>%(LocalPackages.InstallLocation)\packages.%(LocalPackages.PackageName).config</RestoreConfig>
+    </PropertyGroup>
+    
+    <WriteLinesToFile File="$(RestoreConfig)" Lines="@(LocalPackageConfigLine)" Overwrite="true" />
+    <Message Text="Generated $(RestoreConfig) for '%(LocalPackages.PackageName).%(LocalPackages.PackageVersion).nupkg'" />
+    
+    <PropertyGroup>
+      <LocalPackagesSources>@(LocalPackages, ';')</LocalPackagesSources>
+    </PropertyGroup>
+    <Message Text="Generated to package source to be '$(LocalPackages)'"/>
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/Targets/packages.targets]: Reading this message doesn't seem to make since any longer.
